File: packages/react-native-nitro-modules/ios/views/HybridView.swift
//
//  HybridView.swift
//  NitroModules
//
//  Created by Marc Rousavy on 13.01.25.
//

#if canImport(UIKit)

  import Foundation
  import UIKit

  /// A base protocol for all Swift-based Hybrid Views.
  public protocol HybridView: HybridObject {
    /**
     * Get the ``UIView`` this HybridView is holding.
     *
     * This value should not change during the lifetime of this ``HybridView``.
     */
    var view: UIView { get }

    /**
     * Called right before updating props.
     * React props are updated in a single batch/transaction.
     */
    func beforeUpdate()
    /**
     * Called right after updating props.
     * React props are updated in a single batch/transaction.
     */
    func afterUpdate()
  }

  extension HybridView {
    public func beforeUpdate() { /* noop */  }
    public func afterUpdate() { /* noop */  }
  }

  extension HybridView {
    // In a View, this could be the size of the UIView.
    public var memorySize: Int {
      return MemoryHelper.getSizeOf(self.view)
    }
  }

#endif

File: packages/react-native-nitro-modules/android/src/main/java/com/margelo/nitro/views/HybridView.kt
package com.margelo.nitro.views

import android.view.View
import com.margelo.nitro.core.HybridObject

/**
 * A base class for all Kotlin-based Hybrid Views.
 */
abstract class HybridView : HybridObject() {
  /**
   * Get the `View` this HybridView is holding.
   *
   * This value should not change during the lifetime of this `HybridView`.
   */
  abstract val view: View

  /**
   * Called right before updating props.
   * React props are updated in a single batch/transaction.
   */
  open fun beforeUpdate() { /* noop */ }

  /**
   * Called right after updating props.
   * React props are updated in a single batch/transaction.
   */
  open fun afterUpdate() { /* noop */ }

  /**
   * Called when the [HybridView] is about
   * to be dropped and unmounted.
   * This is a good place to clean up view-related
   * resources.
   */
  open fun onDropView() { /* noop */ }
}
